libsolidity/codegen/ContractCompiler.cpp

```diff
@@ -1246,7 +1246,13 @@ bool ContractCompiler::visit(ForStatement const& _forStatement)
 
 	// for's loop expression if existing
 	if (_forStatement.loopExpression())
+	{
+		Arithmetic previousArithmetic = m_context.arithmetic();
+		if (*_forStatement.annotation().isSimpleCounterLoop)
```

Comment:
Maybe we should not set the annotation on the loop but rather on the `i++` expression?

Reply:
Also, as commented elsewhere, this should only be done if optimizations are activated.